packages/shopping-cart/src/state-based-actions.ts

```diff
@@ -0,0 +1,68 @@
+import debugFactory from 'debug';
+import { playQueuedActions } from './shopping-cart-reducer';
+import type { ShoppingCartState, ShoppingCartReducerDispatch, CacheStatus } from './types';
+
+const debug = debugFactory( 'shopping-cart:state-based-actions' );
+
+function fetchInitialCart(
+	state: ShoppingCartState,
+	dispatch: ShoppingCartReducerDispatch,
+	lastCacheStatus: CacheStatus | ''
+): void {
+	const { cacheStatus } = state;
+	if ( cacheStatus === 'fresh' && cacheStatus !== lastCacheStatus ) {
+		debug( 'triggering fetch of initial cart' );
+		dispatch( { type: 'FETCH_INITIAL_RESPONSE_CART' } );
+		dispatch( { type: 'GET_CART_FROM_SERVER' } );
```

Comment:
That's how it used to work, but if a cart reload action was triggered at the wrong time, it could lead to a race condition just like the ones fixed by https://github.com/Automattic/wp-calypso/pull/54300

Reply:
This should be resolved now as I modified the dispatcher to be synchronous (except for the state based actions).